File: medication_timer.rb
# medication_timer.rb
require 'json'
require 'date'
require 'colorize'
require 'optparse'

DATA_FILE = 'medications.json'

class Medication
  attr_accessor :id, :name, :dosage, :interval_hours, :last_taken, :history

  def initialize(name, dosage, interval_hours, id = 0, last_taken = nil, history = [])
    @id = id
    @name = name
    @dosage = dosage
    @interval_hours = interval_hours
    @last_taken = last_taken || Time.now
    @history = history.map { |t| Time.parse(t) }
  end

  def is_due?
    Time.now >= next_dose
  end

  def next_dose
    @last_taken + interval_hours * 3600
  end

  def take_dose
    now = Time.now
    @last_taken = now
    @history << now
  end

  def to_hash
    {
      id: @id,
      name: @name,
      dosage: @dosage,
      interval_hours: @interval_hours,
      last_taken: @last_taken.iso8601,
      history: @history.map(&:iso8601)
    }
  end

  def self.from_hash(h)
    new(h['name'], h['dosage'], h['interval_hours'], h['id'], h['last_taken'], h['history'])
  end
end

class TimerApp
  attr_reader :medications

  def initialize
    @medications = []
    load
  end

  def load
    return unless File.exist?(DATA_FILE)
    data = JSON.parse(File.read(DATA_FILE))
    @medications = data.map { |h| Medication.from_hash(h) }
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@medications.map(&:to_hash)))
  end

  def add_medication(name, dosage, interval)
    max_id = @medications.map(&:id).max || 0
    med = Medication.new(name, dosage, interval, max_id + 1)
    @medications << med
    save
    puts "💊 Added medication: #{name} #{dosage} every #{interval}h (ID: #{med.id})"
  end

  def list
    if @medications.empty?
      puts "No medications."
      return
    end
    puts "\n📋 Medications:"
    @medications.each do |m|
      due = m.is_due? ? "🔴 DUE".red : "⏳ #{m.next_dose.strftime('%H:%M')}".green
      puts "  [#{m.id}] #{m.name} #{m.dosage} – #{due}"
    end
  end

  def take_dose(id)
    med = @medications.find { |m| m.id == id }
    unless med
      puts "❌ Medication with ID #{id} not found."
      return
    end
    med.take_dose
    save
    puts "✅ Took #{med.name} #{med.dosage} at #{Time.now.strftime('%H:%M')}"
  end

  def check_due
    due = @medications.select(&:is_due?)
    if due.empty?
      puts "✅ No medications due right now."
      return
    end
    puts "\n⏰ Due Medications:"
    due.each do |m|
      overdue = ((Time.now - m.next_dose) / 60).to_i
      puts "  [#{m.id}] #{m.name} #{m.dosage} – overdue by #{overdue} min"
    end
  end

  def history
    if @medications.empty?
      puts "No medications."
      return
    end
    puts "\n📜 History (last 5 doses per medication):"
    @medications.each do |m|
      puts "  #{m.name}:"
      recent = m.history.last(5)
      if recent.empty?
        puts "    (no history)"
      else
        recent.each { |t| puts "    - #{t.strftime('%Y-%m-%d %H:%M')}" }
      end
    end
  end
end

options = {}
$command = ARGV.shift
if $command.nil?
  puts "Usage: medication_timer.rb <command> [options]"
  exit 1
end

app = TimerApp.new

case $command
when "add"
  name, dosage, interval = ARGV.shift(3)
  app.add_medication(name, dosage, interval.to_f)
when "list"
  app.list
when "take"
  id = ARGV.shift.to_i
  app.take_dose(id)
when "check"
  app.check_due
when "history"
  app.history
else
  puts "Unknown command."
end

File: medication_timer.js
// medication_timer.js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');
const { program } = require('commander');
const chalk = require('chalk');

const DATA_FILE = 'medications.json';

class Medication {
    constructor(name, dosage, intervalHours, id = 0, lastTaken = null, history = []) {
        this.id = id;
        this.name = name;
        this.dosage = dosage;
        this.intervalHours = intervalHours;
        this.lastTaken = lastTaken ? new Date(lastTaken) : new Date();
        this.history = history.map(d => new Date(d));
    }

    isDue() {
        const next = new Date(this.lastTaken.getTime() + this.intervalHours * 3600000);
        return new Date() >= next;
    }

    nextDose() {
        return new Date(this.lastTaken.getTime() + this.intervalHours * 3600000);
    }

    takeDose() {
        const now = new Date();
        this.lastTaken = now;
        this.history.push(now);
    }

    toJSON() {
        return {
            id: this.id,
            name: this.name,
            dosage: this.dosage,
            intervalHours: this.intervalHours,
            lastTaken: this.lastTaken.toISOString(),
            history: this.history.map(d => d.toISOString())
        };
    }

    static fromJSON(data) {
        return new Medication(data.name, data.dosage, data.intervalHours, data.id, data.lastTaken, data.history);
    }
}

class TimerApp {
    constructor() {
        this.medications = [];
        this.load();
    }

    load() {
        if (fs.existsSync(DATA_FILE)) {
            const data = JSON.parse(fs.readFileSync(DATA_FILE));
            this.medications = data.map(m => Medication.fromJSON(m));
        }
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify(this.medications.map(m => m.toJSON()), null, 2));
    }

    addMedication(name, dosage, interval) {
        const maxId = this.medications.reduce((max, m) => Math.max(max, m.id), 0);
        const med = new Medication(name, dosage, interval, maxId + 1);
        this.medications.push(med);
        this.save();
        console.log(`💊 Added medication: ${name} ${dosage} every ${interval}h (ID: ${med.id})`);
    }

    list() {
        if (this.medications.length === 0) {
            console.log('No medications.');
            return;
        }
        console.log('\n📋 Medications:');
        for (const m of this.medications) {
            const due = m.isDue() ? chalk.red('🔴 DUE') : chalk.green(`⏳ ${m.nextDose().toTimeString().slice(0,5)}`);
            console.log(`  [${m.id}] ${m.name} ${m.dosage} – ${due}`);
        }
    }

    takeDose(id) {
        const med = this.medications.find(m => m.id === id);
        if (!med) {
            console.log(`❌ Medication with ID ${id} not found.`);
            return;
        }
        med.takeDose();
        this.save();
        console.log(`✅ Took ${med.name} ${med.dosage} at ${new Date().toTimeString().slice(0,5)}`);
    }

    checkDue() {
        const due = this.medications.filter(m => m.isDue());
        if (due.length === 0) {
            console.log('✅ No medications due right now.');
            return;
        }
        console.log('\n⏰ Due Medications:');
        for (const m of due) {
            const overdue = Math.floor((new Date() - m.nextDose()) / 60000);
            console.log(`  [${m.id}] ${m.name} ${m.dosage} – overdue by ${overdue} min`);
        }
    }

    history() {
        if (this.medications.length === 0) {
            console.log('No medications.');
            return;
        }
        console.log('\n📜 History (last 5 doses per medication):');
        for (const m of this.medications) {
            console.log(`  ${m.name}:`);
            const recent = m.history.slice(-5);
            if (recent.length === 0) {
                console.log('    (no history)');
            } else {
                for (const d of recent) {
                    console.log(`    - ${d.toISOString().slice(0,16).replace('T', ' ')}`);
                }
            }
        }
    }
}

program
    .command('add <name> <dosage> <interval>')
    .action((name, dosage, interval) => {
        const app = new TimerApp();
        app.addMedication(name, dosage, parseFloat(interval));
    });

program
    .command('list')
    .action(() => {
        const app = new TimerApp();
        app.list();
    });

program
    .command('take <id>')
    .action((id) => {
        const app = new TimerApp();
        app.takeDose(parseInt(id));
    });

program
    .command('check')
    .action(() => {
        const app = new TimerApp();
        app.checkDue();
    });

program
    .command('history')
    .action(() => {
        const app = new TimerApp();
        app.history();
    });

program.parse(process.argv);
